### ISSUE TITLE
Infinite recursion in spack spec

### ISSUE THREAD

**Issue by adamjstewart** [spack/spack#2495]
When I clone a fresh copy of Spack and run:
```
$ spack spec libxext
Input spec
--------------------------------
libxext

Normalized
--------------------------------
libxext
    ^libx11@1.6:
        ^inputproto
            ^pkg-config@0.9.0:
            ^util-macros
        ^kbproto
        ^libxcb@1.1.92:
            ^libpthread-stubs
            ^libxau@0.99.2:
                ^xproto@7.0.17:
            ^libxdmcp
            ^python@2:2.8
                ^bzip2
                ^ncurses
                ^openssl
                    ^zlib
                ^readline
                ^sqlite
            ^xcb-proto
        ^xextproto@7.1.99:
        ^xtrans

Concretized
--------------------------------
Exception RuntimeError: 'maximum recursion depth exceeded while calling a Python object' in <type 'exceptions.TypeError'> ignored
Exception RuntimeError: 'maximum recursion depth exceeded while calling a Python object' in <type 'exceptions.TypeError'> ignored
Traceback (most recent call last):
  File "./spack", line 198, in <module>
    main()
  File "./spack", line 175, in main
    return_val = command(parser, args)
  File "/blues/gpfs/home/ajstewart/spack/lib/spack/spack/cmd/spec.py", line 78, in spec
    spec.concretize()
  File "/blues/gpfs/home/ajstewart/spack/lib/spack/spack/spec.py", line 1507, in concretize
    changes = (self.normalize(force),
  File "/blues/gpfs/home/ajstewart/spack/lib/spack/spack/spec.py", line 1809, in normalize
    any_change = self._normalize_helper(visited, spec_deps, provider_index)
  File "/blues/gpfs/home/ajstewart/spack/lib/spack/spack/spec.py", line 1763, in _normalize_helper
    pkg_dep, deptypes, visited, spec_deps, provider_index)
  File "/blues/gpfs/home/ajstewart/spack/lib/spack/spack/spec.py", line 1734, in _merge_dependency
    visited, spec_deps, provider_index)
  File "/blues/gpfs/home/ajstewart/spack/lib/spack/spack/spec.py", line 1763, in _normalize_helper
    pkg_dep, deptypes, visited, spec_deps, provider_index)
  File "/blues/gpfs/home/ajstewart/spack/lib/spack/spack/spec.py", line 1734, in _merge_dependency
    visited, spec_deps, provider_index)
  File "/blues/gpfs/home/ajstewart/spack/lib/spack/spack/spec.py", line 1763, in _normalize_helper
    pkg_dep, deptypes, visited, spec_deps, provider_index)
  File "/blues/gpfs/home/ajstewart/spack/lib/spack/spack/spec.py", line 1734, in _merge_dependency
    visited, spec_deps, provider_index)
  File "/blues/gpfs/home/ajstewart/spack/lib/spack/spack/spec.py", line 1758, in _normalize_helper
    pkg_dep = self._evaluate_dependency_conditions(dep_name)
  File "/blues/gpfs/home/ajstewart/spack/lib/spack/spack/spec.py", line 1629, in _evaluate_dependency_conditions
    sat = self.satisfies(when_spec, strict=True)
  File "/blues/gpfs/home/ajstewart/spack/lib/spack/spack/spec.py", line 2058, in satisfies
    return self.satisfies_dependencies(other, strict=deps_strict)
  File "/blues/gpfs/home/ajstewart/spack/lib/spack/spack/spec.py", line 2087, in satisfies_dependencies
    self_index = ProviderIndex(self.traverse(), restrict=True)
  File "/blues/gpfs/home/ajstewart/spack/lib/spack/spack/provider_index.py", line 87, in __init__
    self.update(spec)
  File "/blues/gpfs/home/ajstewart/spack/lib/spack/spack/provider_index.py", line 99, in update
    pkg = spec.package
  File "/blues/gpfs/home/ajstewart/spack/lib/spack/spack/spec.py", line 945, in package
    return spack.repo.get(self)
  File "/blues/gpfs/home/ajstewart/spack/lib/spack/spack/repository.py", line 75, in converter
    return function(self, spec_like, *args, **kwargs)
  File "/blues/gpfs/home/ajstewart/spack/lib/spack/spack/repository.py", line 318, in get
    return self.repo_for_pkg(spec).get(spec)
  File "/blues/gpfs/home/ajstewart/spack/lib/spack/spack/repository.py", line 75, in converter
    return function(self, spec_like, *args, **kwargs)
  File "/blues/gpfs/home/ajstewart/spack/lib/spack/spack/repository.py", line 567, in get
    key = hash(spec)
  File "/blues/gpfs/home/ajstewart/spack/lib/spack/llnl/util/lang.py", line 254, in <lambda>
    setter('__hash__', lambda self: hash(self._cmp_key()))
  File "/blues/gpfs/home/ajstewart/spack/lib/spack/spack/spec.py", line 2330, in _cmp_key
    for name, d in sorted(self._dependencies.items()))
  File "/blues/gpfs/home/ajstewart/spack/lib/spack/spack/spec.py", line 2330, in <genexpr>
    for name, d in sorted(self._dependencies.items()))
...
  File "/blues/gpfs/home/ajstewart/spack/lib/spack/spack/spec.py", line 2330, in _cmp_key
    for name, d in sorted(self._dependencies.items()))
  File "/blues/gpfs/home/ajstewart/spack/lib/spack/spack/spec.py", line 2330, in <genexpr>
    for name, d in sorted(self._dependencies.items()))
  File "/blues/gpfs/home/ajstewart/spack/lib/spack/llnl/util/lang.py", line 254, in <lambda>
    setter('__hash__', lambda self: hash(self._cmp_key()))
  File "/blues/gpfs/home/ajstewart/spack/lib/spack/spack/spec.py", line 2330, in _cmp_key
    for name, d in sorted(self._dependencies.items()))
TypeError: object of type 'generator' has no len()
```
This also occurs for me with other very large packages like `dealii`. Interestingly enough, I am unable to reproduce this on another computer. I've tried newer versions of Python but no luck. This definitely wasn't a problem until recently. Any ideas?

**Comment by alalazo** [comment]
Can you post the output of this where the problem occurs ?
```console
$ python
Python 2.7.6 (default, Oct 26 2016, 20:30:19) 
[GCC 4.8.4] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> sys.getrecursionlimit()
1000
```

**Comment by adamjstewart** [comment]
Looks like the limit is 1000 on both computers, so this isn't the problem.

**Comment by adamjstewart** [comment]
Identified the difference between computers. On one computer, I have the following in my `packages.yaml`:
```yaml
packages:
  python:
    variants: +tk
```
the other does not. When I copy this to the other computer, it breaks all of a sudden.

**Comment by adamjstewart** [comment]
So either adding the tk dependencies to Python just barely pushes me past the 1,000 recursion depth limit, or it is introducing an infinite dependency loop somewhere.

**Comment by adamjstewart** [comment]
For the record, anyone should be able to reproduce this problem. The following works:
```
$ spack spec libxext ^python~tk
```
The following crashes:
```
$ spack spec libxext ^python+tk
```

**Comment by alalazo** [comment]
> So either adding the tk dependencies to Python just barely pushes me past the 1,000 recursion depth limit, or it is introducing an infinite dependency loop somewhere.

This is easy to check tweaking `spack`. Just add:
```
import sys

sys.setrecursionlimit(<the-number-you-want>)
```
at the beginning of the script and see if it solves the problem.

**Comment by adamjstewart** [comment]
@alalazo I tried setting the recursion limit to 100,000 and it segfaulted. Sounds like an infinite loop to me.

**Comment by alalazo** [comment]
It's an infinite recursion:
```console
  File "/home/mculpo/github/spack-llnl/lib/spack/llnl/util/lang.py", line 254, in <lambda>
    setter('__hash__', lambda self: hash(self._cmp_key()))
  File "/home/mculpo/github/spack-llnl/lib/spack/spack/spec.py", line 2330, in _cmp_key
    for name, d in sorted(self._dependencies.items()))
  File "/home/mculpo/github/spack-llnl/lib/spack/spack/spec.py", line 2330, in <genexpr>
    for name, d in sorted(self._dependencies.items()))
  File "/home/mculpo/github/spack-llnl/lib/spack/llnl/util/lang.py", line 254, in <lambda>
    setter('__hash__', lambda self: hash(self._cmp_key()))
```

**Comment by alalazo** [comment]
Gotcha! Not a dag anymore:
```console
$ spack spec tk
Input spec
--------------------------------
tk

Normalized
--------------------------------
tk
    ^tcl
        ^zlib

Concretized
--------------------------------
tk@8.6.5%gcc@4.8+X arch=linux-ubuntu14-x86_64
    ^libx11@1.6.3%gcc@4.8 arch=linux-ubuntu14-x86_64
        ^inputproto@2.3.2%gcc@4.8 arch=linux-ubuntu14-x86_64
            ^pkg-config@0.29.1%gcc@4.8+internal_glib arch=linux-ubuntu14-x86_64
            ^util-macros@1.19.0%gcc@4.8 arch=linux-ubuntu14-x86_64
        ^kbproto@1.0.7%gcc@4.8 arch=linux-ubuntu14-x86_64
        ^libxcb@1.12%gcc@4.8 arch=linux-ubuntu14-x86_64
            ^libpthread-stubs@0.3%gcc@4.8 arch=linux-ubuntu14-x86_64
            ^libxau@1.0.8%gcc@4.8 arch=linux-ubuntu14-x86_64
                ^xproto@7.0.29%gcc@4.8 arch=linux-ubuntu14-x86_64
            ^libxdmcp@1.1.2%gcc@4.8 arch=linux-ubuntu14-x86_64
            ^python@2.7.12%gcc@4.8~tk~ucs4 arch=linux-ubuntu14-x86_64
                ^bzip2@1.0.6%gcc@4.8 arch=linux-ubuntu14-x86_64
                ^ncurses@6.0%gcc@4.8 arch=linux-ubuntu14-x86_64
                ^openssl@1.0.2j%gcc@4.8 arch=linux-ubuntu14-x86_64
                    ^zlib@1.2.8%gcc@4.8+pic arch=linux-ubuntu14-x86_64
                ^readline@6.3%gcc@4.8 arch=linux-ubuntu14-x86_64
                ^sqlite@3.8.5%gcc@4.8 arch=linux-ubuntu14-x86_64
            ^xcb-proto@1.12%gcc@4.8 arch=linux-ubuntu14-x86_64
        ^xextproto@7.3.0%gcc@4.8 arch=linux-ubuntu14-x86_64
        ^xtrans@1.3.5%gcc@4.8 arch=linux-ubuntu14-x86_64
    ^tcl@8.6.5%gcc@4.8 arch=linux-ubuntu14-x86_64
```

**Comment by adamjstewart** [comment]
@alalazo Is it just because `python+tk` depends on `tk`, which depends on `python+tk`?

**Comment by adamjstewart** [comment]
If I remove the Python dependencies from `libxcb` and `xcb-proto`, things work again.

**Comment by adamjstewart** [comment]
Looks like this problem arose after #2053 was merged. If I comment out the tk dependency on libx11, it works again.

**Comment by citibeth** [comment]
This problem (circular dependency) was also noted in #2098 and is similar to #2321 (where we discussed a bunch of ideas, resulting in #2325)

We will see this problem more as we try to make Spack more complete.  I suggest the following approaches to it:

1. The theoretically correct solution to this problem is to recognize that dependencies exist as a SPIRAL, not a DAG.  At the most basic level... if you want anything, you need to build a compiler.  But to get that compiler, you need a bunch of other stuff.  Real systems are bootstrapped over time to incrementally increasing versions.  If we lost all binary copies of GCC (and other compilers), someone have to re-do that process.

Spack currently exists in a comfortable subset of this worldwide software spiral, in which it insists that dependencies are a DAG, not a spiral.  If Spack could be extended somewhat, this problem could be solved in theory.  In this case, we would want `python+tk` --> `libxcb` --> `python`.  Right now, Spack can only have `python` once in a DAG, so this dependency relationship is not possible.

2. So far, this "dependency spiral" problem has happened because some package needs Python2 as a build dependency.  Given that Spack is *by definition* running on a Python2 installation, it should NEVER need to build Python to satisfy a Python2 build dependency.  Therefore, it could choose to satisfy such dependencies simply by querying the Python it's running.  I know this sounds hackey.  But it would deliver some real benefits in practical use: (a) eliminate all circular dependencies we've seen so far, (b) eliminate the "need" to build an additional Python2 just for a build dependency.

3. One can always specify certain things as foreign packages in `packages.yaml` --- for example, `libxcb`, `python@2.7`, etc.  I know this isn't really a solution, more of a workaround for now.








2. Use Spack python

**Comment by adamjstewart** [comment]
This problem happened to me again for `libxml2` and anything that depends on it because I have `libxml2+python` set in my `packages.yaml`.

**Comment by scheibelp** [comment]
https://github.com/LLNL/spack/pull/2548 is an attempt to resolve this

**Comment by alalazo** [comment]
I don't see the infinite recursion anymore, at least in the specs above.

**Comment by adamjstewart** [comment]
That's because we removed a couple Python build dependencies to prevent it. See #4145.

**Comment by scheibelp** [comment]
@ax3l at first glance this error appears to be different: the issue as reported in the initial description is triggered by concretization (from the `spack spec` command); this is triggered by the environment script, and (based on the message itself) config parsing.

If you agree, could you open a new issue, which highlights the specific part that gets repeated:

```
  File "/home/axel/src/spack/lib/spack/llnl/util/lang.py", line 556, in __getattr__
    return getattr(self.instance, name)
  [Previous line repeated 318 more times]
```

(in which case I will collapse this and https://github.com/spack/spack/issues/2495#issuecomment-419636297)

**Comment by ax3l** [comment]
@alalazo thank you for trying. I guess it's something unclean in either my configs or source tree. Can not reproduce it in Docker. But I can't run `spack clean -a` either, since I can't source it xD I checked out a fresh spack and removed my configs and still have the issue. Weird.

**Comment by healther** [comment]
@ax3l can you try deleting all `.pyc` files?

**Comment by ax3l** [comment]
@healther did so as well. and all `__pycache__` dirs

**Comment by alalazo** [comment]
Closing since the feature request that would solve this issue (circular dependencies in the DAG) is tracked in #2498
